flow/record/adapter/gcs.py

```diff
@@ -0,0 +1,97 @@
+import logging
+from fnmatch import fnmatch
+from typing import Iterator, Union
+
+from google.cloud.storage.client import Client
+from google.cloud.storage.fileio import BlobReader, BlobWriter
+
+from flow.record.adapter import AbstractReader, AbstractWriter
+from flow.record.base import Record, RecordAdapter
+from flow.record.selector import CompiledSelector, Selector
+
+__usage__ = """
+Google Cloud Storage adapter
+---
+Read usage: rdump gcs://[PROJECT-ID]:[BUCKET-ID]?path=[PATH]
+[PROJECT-ID]: Google Cloud Project ID
+[BUCKET-ID]: Bucket ID
+[path]: Path to look for files, with support for glob-pattern matching
+
+Write usage: rdump gcs://[PROJECT-ID]:[BUCKET-ID]?path=[PATH]
+[PROJECT-ID]: Google Cloud Project ID
+[BUCKET-ID]: Bucket ID
+[path]: Path to write records to
+"""
+
+log = logging.getLogger(__name__)
+
+GLOB_CHARACTERS = "*?[]"
+
+
+class GcsReader(AbstractReader):
+    def __init__(self, uri: str, path: str, selector: Union[None, Selector, CompiledSelector] = None, **kwargs) -> None:
+        self.selector = selector
+        project_name = uri[: uri.find(":")]
+        bucket_name = uri[uri.find(":") + 1 :]
+
+        self.gcs = Client(project=project_name)
+        self.bucket = self.gcs.bucket(bucket_name)
+
+        # GCS Doesn't support iterating blobs using a glob pattern, so we have to do that ourselves.
+        # To split the path prefix from the glob-specific stuff, we have to find the first place where
+        # the glob starts. We'll then go through all files that match the path prefix before the glob,
+        # and do fnmatch ourselves to check whether any given blob matches with our glob.
+        lowest_pos = min([path.find(char) if path.find(char) >= 0 else float("inf") for char in GLOB_CHARACTERS])
```

Comment:
i think you can also use `partition()` here to make the code more concise. Also, does this support multiple glob patterns?

Reply:
As far as I could tell, `partition()` isn't really suitable for choosing the first occurence out of multiple seperator characters. To make the code a bit more readable, I opted to feed `min()` an iterator of glob characters and defaulting to `-1` if none matched. 

This PR uses `fnmatch` to verify whether or not the path matches with the pattern, and you can use multiple glob characters in your matching:

```
In [1]: from fnmatch import fnmatch

In [2]: pattern = "*/*/*/t?st.records"

In [3]: fnmatch("a/b/c/tst.records", pattern)
Out[3]: False

In [4]: fnmatch("a/b/c/test.records", pattern)
Out[4]: True
```

Is this what you mean with supporting multiple glob patterns?

Reply:
In `dissect.target` we use a regex for this:
```python
re_glob_index = re.compile(r"(?<=\/)[^\/]*[*?[]\/?")


def glob_split(pattern: str, alt_separator: str = "") -> str:
    # re_glob_index expects a normalized pattern
    pattern = normalize(pattern, alt_separator=alt_separator)

    first_glob = re_glob_index.search(pattern)

    if not first_glob:
        return pattern, ""

    pos = first_glob.start()
    return pattern[:pos], pattern[pos:]
```

Reply:
You can also use `re.split` in combination with `maxsplit`, for example:

```python
>>> path = "/data/lake/*/citrix/????"
>>> parts = re.split("[\?\*\[\]]", path, maxsplit=1)
>>> print(parts)
['/data/lake/', '/citrix/????']

>>> parts = re.split("[\?\*\[\]]", "/data/lake/test", maxsplit=1)
>>> print(parts)
['/data/lake/test']
```